test/test_service.rb

```diff
@@ -273,4 +273,66 @@ def test_patch_service
       data['metadata']['annotations']['key'] == 'value'
     end
   end
+
+  def test_json_patch_service
+    service = Kubeclient::Resource.new
+    name = 'my_service'
+
+    service.metadata = {}
+    service.metadata.name      = name
+    service.metadata.namespace = 'development'
+
+    stub_core_api_list
+    expected_url = "http://localhost:8080/api/v1/namespaces/development/services/#{name}"
+    stub_request(:patch, expected_url)
+      .to_return(body: open_test_file('service_patch.json'), status: 200)
+
+    patch = {
+      metadata: {
+        annotations: {
+          key: 'value'
+        }
+      }
+    }
+
+    client = Kubeclient::Client.new('http://localhost:8080/api/', 'v1')
+    service = client.json_patch_service(name, patch, 'development')
+    assert_kind_of(RecursiveOpenStruct, service)
+
+    assert_requested(:patch, expected_url, times: 1) do |req|
+      data = JSON.parse(req.body)
+      data['metadata']['annotations']['key'] == 'value'
+    end
+  end
+
+  def test_merge_patch_service
+    service = Kubeclient::Resource.new
+    name = 'my_service'
+
+    service.metadata = {}
+    service.metadata.name      = name
+    service.metadata.namespace = 'development'
+
+    stub_core_api_list
+    expected_url = "http://localhost:8080/api/v1/namespaces/development/services/#{name}"
+    stub_request(:patch, expected_url)
+      .to_return(body: open_test_file('service_patch.json'), status: 200)
+
+    patch = {
+      metadata: {
+        annotations: {
+          key: 'value'
+        }
+      }
+    }
```

Comment:
(This one can stay same as strategic `patch` :+1:
When arrays are involved *result* might differ but we're testing kubeclient not k8s here, expressing this difference would require extra effort with mock results but won't make the test any stronger.)

Reply:
>but we're testing kubeclient not k8s here

In this case the line feels a little blurry, since I'm quite interested that the response being sent back by the server is valid. It looks like `test_patch_service` is completely stubbing out the `patch` call. In fact, I can remove the `content-type` header from the code and `test_patch_service` will still pass.

One approach I could take is to create a fixture by executing the code against a cluster so at least we have a valid snapshot. I see that VCR is used in some places in this repo but understand that it has its own drawbacks as well.

I'm interested to hear your thoughts here. In the worst case I'll get a working test that won't fail if a regression is introduced, since we're completely stubbing that part out. On the other hand, it does at least show that the client is discovering the appropriate method and making it available on resources 🤷‍♂️

Reply:
> I can remove the content-type header from the code and test_patch_service will still pass.

Oh, something ate my comment with suggestions for testing Content-type.  (untested, only googled)
One way is add after the `stub_request`:
```
  .with(headers: { 'Content-Type' => '...'})
```
another is extend the condition in `assert_requested`:
```
  req.headers['Content-Type'] == '...' &&
  data[...] == ...
```